src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -16,33 +16,38 @@
 
 #include "ray/core_worker/transport/dependency_resolver.h"
 #include "ray/gcs/pb_util.h"
+#include "ray/stats/metric_defs.h"
 
 namespace ray {
 namespace core {
 
-Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
-  RAY_LOG(DEBUG) << "Submit task " << task_spec.TaskId();
+Status CoreWorkerDirectTaskSubmitter::SubmitTask(
+    std::shared_ptr<TaskSpecification> task_spec) {
```

Comment:
hmmm so the problem is that the closure we pass into `ResolveDependencies` is capture-copying `task_spec`, which makes it immutable (and we want to mutate it). 

In my mind, the solutions seems to be:
1. pass by reference instead of copy (but `task_spec` is a stack variable and the lambda outlives it, hence the shared_ptr)
2. Make the lambda `mutable` which works but feels like an ugly bug just waiting to happen (since the lambda will be mutating a copy of the task spec, not the original).

I don't have strong opinions on this though, so let me know if you prefer another implementation.

Reply:
done